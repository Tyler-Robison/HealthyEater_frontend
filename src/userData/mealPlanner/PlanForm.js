import React, { useContext } from "react";
import { useFormik } from "formik";
import PlannerAPI from "../../APIs/plannerAPI";
import GlobalContext from "../../context/GlobalContext";

const PlanForm = ({ days }) => {

    const { token, currentUser, setCurrentUser } = useContext(GlobalContext);

    const formik = useFormik({
        initialValues: {
            daySelect: '',
            mealSelect: ''
        },
        onSubmit: values => handleFormikSubmit(values)
    })

    const handleFormikSubmit = async (values) => {
        let { daySelect, mealSelect } = values

        const meal = (mealSelect === '' ? currentUser.recipes[0] : 
        JSON.parse(mealSelect))

        if (daySelect === '') daySelect = 'Mon'
        const data = {
            day: daySelect,
            recipe_id: meal.recipe_id
        }

        const mealRes = await PlannerAPI.setMeal(currentUser.id, token, data)
        // console.log('meal res', mealRes)

        // mealPlannerRow contains:
        // id, recipe_id, day from user_mealplan table AND
        // ww_points, name from recipes table 
        currentUser.mealplan.push(mealRes.mealplannerRow)

        setCurrentUser({ ...currentUser, currentUser })
    }

    const dayValues = days.map((day, idx) => {
        return <option key={idx} value={day}>{day}</option>
    })

    const recipeValues = currentUser.recipes.map(recipe => {
        const recipeObj = {
            name: recipe.name,
            recipe_id: recipe.recipe_id,
            ww_points: recipe.ww_points
        }
        
        // store recipeObj as string, convert back to object upon formSubmit
        return <option key={recipe.recipe_id} value={JSON.stringify(recipeObj)}>{recipe.name}</option>
    })

    return (
        <form onSubmit={formik.handleSubmit}>
            <label htmlFor="daySelect">Select Day</label>
            <select
                id="daySelect"
                name="daySelect"
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
            >{dayValues}</select>

            <label htmlFor="mealSelect">Select Meal</label>
            <select
                id="mealSelect"
                name="mealSelect"
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
            >{recipeValues}</select>


            <button className="general-btn" type="submit">Add to Calendar</button>
        </form>
    )
}

export default PlanForm;