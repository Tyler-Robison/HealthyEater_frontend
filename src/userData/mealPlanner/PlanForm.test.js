import { render, screen } from '@testing-library/react';
import PlanForm from './PlanForm';
import { MemoryRouter } from "react-router-dom";
import ContextProvider from '../../testContext';

const days = ['Mon', 'Tues', 'Wed', 'Thurs', 'Fri', 'Sat', 'Sun'];

it("renders without crashing", function () {
    <MemoryRouter>
        <PlanForm days={days} />
    </MemoryRouter>
});


// daySelect must correctly display values in days array
// mealSelect must correctly display recipe values from currentUser.recipes
it("matches snapshot", function () {

    const container = render(
        <MemoryRouter>
            <ContextProvider >
                <PlanForm days={days} />
            </ContextProvider>
        </MemoryRouter>
    );
    expect(container.asFragment()).toMatchSnapshot();
});