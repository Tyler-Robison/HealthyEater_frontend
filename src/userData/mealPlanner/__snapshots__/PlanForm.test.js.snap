// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MealPlanner-form"
  >
    <form>
      <label
        for="daySelect"
      >
        Select Day
      </label>
      <select
        id="daySelect"
        name="daySelect"
      >
        <option
          value="Mon"
        >
          Mon
        </option>
        <option
          value="Tues"
        >
          Tues
        </option>
        <option
          value="Wed"
        >
          Wed
        </option>
        <option
          value="Thurs"
        >
          Thurs
        </option>
        <option
          value="Fri"
        >
          Fri
        </option>
        <option
          value="Sat"
        >
          Sat
        </option>
        <option
          value="Sun"
        >
          Sun
        </option>
      </select>
      <label
        for="mealSelect"
      >
        Select Meal
      </label>
      <select
        id="mealSelect"
        name="mealSelect"
      >
        <option
          value="{\\"name\\":\\"bacon\\",\\"recipe_id\\":1111,\\"ww_points\\":20}"
        >
          bacon
        </option>
        <option
          value="{\\"name\\":\\"eggs\\",\\"recipe_id\\":2222,\\"ww_points\\":10}"
        >
          eggs
        </option>
        <option
          value="{\\"name\\":\\"salad\\",\\"recipe_id\\":3333,\\"ww_points\\":5}"
        >
          salad
        </option>
        <option
          value="{\\"name\\":\\"bread\\",\\"recipe_id\\":4444,\\"ww_points\\":7}"
        >
          bread
        </option>
      </select>
      <button
        class="general-btn"
        type="submit"
      >
        Add to Calendar
      </button>
    </form>
  </div>
</DocumentFragment>
`;
